### PR TITLE
Add automated test for tcms.kiwi_auth.views

### LINKED ISSUE #1608: Automated tests for missing coverage in tcms.kiwi_auth.views
Kiwi TCMS needs additional automated test(s) as mentioned in subject.

**Definition of Done**:
- full coverage for the module or class mentioned in the title. This means both positive/negative scenarios where this makes sense or scenarios which traverse different branches of the code where that makes sense.
- test class/test method names are descriptive and reflect the behavior of the subject under test.  A good scheme to follow is `Test<ClassUnderTest>.test_[who][what][expected_results]`
- CI pass
- All code review passed & PR merged

Hints:
- Coverage report is at https://codecov.io/gh/kiwitcms/Kiwi/tree/master/tcms
- Devel getting started docs: https://kiwitcms.readthedocs.io/en/latest/contribution.html#testing
- Django testing documentation: https://docs.djangoproject.com/en/3.0/topics/testing/

*This issue is part of Kiwi TCMS open source bounty program. For more information see the link(s) in [bounty-program milestone](https://github.com/kiwitcms/Kiwi/milestone/4)*

### PR DESCRIPTION
**Pull request by iamabhishek0**
Fix #1608

### CODE REVIEW

File: tcms/kiwi_auth/tests/test_views.py [hunk 1]
```diff
@@ -199,6 +201,20 @@ def test_register_user_and_activate_by_admin(self):
...
+            self.register_url,
+            {
+                "username": "kiwi-tester",
+                "password1": "password-1",
+                "password2": "password-2",
+                "email": "new-tester@example.com",
+            },
+            follow=False,
+        )
+        self.assertEqual(response.status_code, 200)
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_views.py:215]: there should be an error message shown telling the user which field is invalid/has error. Assert on it.

File: tcms/kiwi_auth/tests/test_views.py [hunk 2]
```diff
@@ -262,3 +278,49 @@ def test_confirm(self):
...
+    @classmethod
+    def setUpTestData(cls):
+        cls.tester = UserFactory()
+        cls.tester.set_password("password")
+        cls.tester.save()
+
+    def test_view_if_user_is_valid(self):
+        self.client.login(  # nosec:B106:hardcoded_password_funcarg
+            username=self.tester.username, password="password"
+        )
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_views.py:318]: LoggedInTestCase does the login & setup for you already.

File: tcms/kiwi_auth/tests/test_views.py [hunk 3]
```diff
@@ -262,3 +278,49 @@ def test_confirm(self):
...
+        response = self.client.get(
+            reverse("tcms-profile", args=[self.tester.username]), follow=True
+        )
+        self.assertRedirects(response, "/admin/auth/user/2/change/", status_code=302)
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_views.py:322]: this is covered in test_profiles.py

File: tcms/kiwi_auth/tests/test_views.py [hunk 4]
```diff
@@ -262,3 +278,49 @@ def test_confirm(self):
...
+
+    def test_view_if_user_is_invalid(self):
+        response = self.client.get(reverse("tcms-profile", args=["kiwi-tester"]))
+        self.assertEqual(response.status_code, 404)
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_views.py:326]: this isn't so you can merge it with test_profiles.py and adjust the existing test class to reuse LoggedInTestCase

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/kiwi_auth/tests/test_views.py [hunk 5]
```diff
@@ -199,6 +201,25 @@ def test_register_user_and_activate_by_admin(self):
...
+                "password2": "password-2",
+                "email": "new-tester@example.com",
+            },
+            follow=False,
+        )
+
+        self.assertEqual(
+            response.context['form'].error_messages,
+            {'password_mismatch': 'The two password fields didn’t match.'}
+        )
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_views.py:219]: It is better to use `assertContains()` which makes sure that the page is actually showing the message.

File: tcms/kiwi_auth/tests/test_views.py [hunk 6]
```diff
@@ -262,3 +283,26 @@ def test_confirm(self):
...
+    def test_get_template_names(self):
+        try:
+            response = self.client.get(reverse("tcms-login"))
+            self.assertIsNotNone(response.template_name)
+            self.assertEqual(
+                response.template_name,
+                ["registration/custom_login.html", "registration/login.html"],
+            )
+        except TemplateDoesNotExist:
+            self.fail("Failed to load template")
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_views.py:300]: Why do we need a try-except here ? From what I can tell if there's an uncaught exception the test will report FAIL/ERROR. OTOH if the templates don't match then the assertion will fail. 

I can't figure out in which case you want to catch a TemplateDoesNotExist exception and then report a failed test and that would be a good thing.

File: tcms/kiwi_auth/tests/test_profiles.py [hunk 7]
```diff
@@ -3,29 +3,22 @@
...
+from tcms.tests import create_request_user, LoggedInTestCase
 from tcms.tests import user_should_have_perm
 
 
-class TestProfilesView(TestCase):
+class TestProfilesView(LoggedInTestCase):
     """Test the profiles view functionality"""
     @classmethod
     def setUpClass(cls):
         super(TestProfilesView, cls).setUpClass()
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_profiles.py:17]: You can also use Python 3 super -> `super().setUpClass()`

File: tcms/kiwi_auth/tests/test_profiles.py [hunk 8]
```diff
@@ -35,26 +28,20 @@ def test_user_can_view_their_own_profile(self):
...
-
         url = reverse('tcms-profile', args=[self.somebody_else.username])
         response = self.client.get(url, follow=True)
 
         self.assertContains(response, self.somebody_else.username)
         self.assertContains(response, self.somebody_else.email)
         self.assertNotContains(response, 'name="_save"')
+
+    def test_view_if_user_is_invalid(self):
+        response = self.client.get(reverse("tcms-profile", args=["kiwi-tester"]))
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_profiles.py:46]: "kiwi-tester" sounds like something that somebody may use in the future. Better use a more verbose username like "non-existing-username" or similar.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Review by atodorov — CHANGES_REQUESTED** [review]
Squash the 2 commits together.

File: tcms/kiwi_auth/tests/test_views.py (hunk 6 above)
Comment (iamabhishek0) [reply tcms/kiwi_auth/tests/test_views.py]: I thought of considering the case when templates don't match, in that case TemplateDoesNotExist is raised. But now I realize that if the templates don't match, then the assertion will fail, so adding try-except here don't make sense. I have removed it now.

File: tcms/kiwi_auth/tests/test_views.py [hunk 9]
```diff
@@ -199,6 +200,24 @@ def test_register_user_and_activate_by_admin(self):
...
+                "password1": "password-1",
+                "password2": "password-2",
+                "email": "new-tester@example.com",
+            },
+            follow=False,
+        )
+
+        self.assertContains(
+            response, 'The two password fields didn’t match.'
+        )
```
Comment (atodorov) [inline tcms/kiwi_auth/tests/test_views.py:217]: Tests in French are failing b/c this error message appears translated to the user when running in a different locale. 
https://github.com/kiwitcms/Kiwi/pull/2143/checks?check_run_id=1537516256

When asserting on it you should assert on the translation. See how other tests are doing it.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Review by atodorov — APPROVED** [review]

**Pull request merged**